torch/csrc/autograd/custom_function.h

```diff
@@ -11,4 +11,197 @@ TORCH_API variable_list _wrap_outputs(
+  std::unordered_set<at::TensorImpl*> dirty_inputs;
+};
+
+struct TORCH_API VariableInfo {
+  explicit VariableInfo(const Variable& var);
+
+  Variable zeros(at::OptionalDeviceGuard& device_guard) const;
+
+  at::Backend backend = at::Backend::Undefined;
+  at::Device device = at::kCPU;
+  at::ScalarType scalar_type = at::kFloat;
+  std::vector<int64_t> size;
+  bool requires_grad;
+};
+
+template <class T>
+struct CppNode : public Node {
+
+  variable_list apply(variable_list&& inputs) override;
+  AutogradContext ctx;
+  std::vector<bool> is_variable_input;
+  std::vector<VariableInfo> input_info;
+  std::vector<VariableInfo> output_info;
+
+  void release_variables() override;
+};
+
+template <typename T>
+using enable_if_var_t = typename std::enable_if<std::is_constructible<Variable, T>::value>::type;
+
+template <typename T>
+using enable_if_not_var_t = typename std::enable_if<!std::is_constructible<Variable, T>::value>::type;
+
+template <typename T, typename... Args>
+enable_if_not_var_t<T> extract_vars(std::vector<bool> &is_var, variable_list& list, T&& cur, Args&& ... args) {
+  is_var.push_back(false);
+  extract_vars(is_var, list, std::forward<Args>(args)...);
+}
+
+template <typename T, typename... Args>
+enable_if_var_t<T> extract_vars(std::vector<bool> &is_var, variable_list& list, T&& cur, Args&& ... args) {
+  is_var.push_back(true);
+  list.emplace_back(cur);
+  extract_vars(is_var, list, std::forward<Args>(args)...);
+}
+
+template <typename... Args>
+void extract_vars(std::vector<bool> &is_var, variable_list& list, Args&& ... args) {
+}
+
+template<class T>
+template<typename... Args>
+variable_list Function<T>::apply(Args&&... args) {
+  std::shared_ptr<CppNode<T>> node(new CppNode<T>, deleteNode);
+  variable_list input_vars;
+
+  const size_t num_inputs = sizeof...(Args);
+  input_vars.reserve(num_inputs);
+  node->is_variable_input.reserve(num_inputs);
+
+  extract_vars(node->is_variable_input, input_vars, args...);
+
+  bool is_executable =  GradMode::is_enabled() && any_variable_requires_grad(input_vars);
+  auto next_edges = collect_next_edges(input_vars);
+  set_ctx_grad_fn(node->ctx, node);
+  node->set_next_edges(std::move(next_edges));
+  node->clear_input_metadata();
+
+  node->input_info.reserve(input_vars.size());
+  for (auto& var : input_vars) {
+      node->input_info.emplace_back(var);
+  }
+
+  variable_list outputs;
+  {
+    AutoGradMode grad_mode(false);
+    outputs = T::forward(&node->ctx, std::forward<Args>(args)...);
+  }
+
+  auto wrapped_outputs = _wrap_outputs(input_vars, node->ctx.get_non_differentiable(), node->ctx.get_dirty(), outputs, is_executable ? node : nullptr);
+
+  node->output_info.reserve(wrapped_outputs.size());
+  for (auto& output : wrapped_outputs) {
+    if (is_executable) {
+      node->output_info.emplace_back(output);
+    }
+  }
+
+  return wrapped_outputs;
+}
+
+template<class T>
+variable_list CppNode<T>::apply(variable_list&& inputs) {
```

Comment:
I'd recommend commenting here that this logic is the same as in `PyNode::apply` and so if you modify here, you should modify PyNode too. (And vice versa!)